src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs

```diff
@@ -0,0 +1,574 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Serializes BuildEventArgs-derived objects into a provided BinaryWriter
+    /// </summary>
+    internal class BuildEventArgsWriter
+    {
+        private readonly BinaryWriter binaryWriter;
+
+        /// <summary>
+        /// Initializes a new instance of BuildEventArgsWriter with a BinaryWriter
+        /// </summary>
+        /// <param name="binaryWriter">A BinaryWriter to write the BuildEventArgs instances to</param>
+        public BuildEventArgsWriter(BinaryWriter binaryWriter)
+        {
+            this.binaryWriter = binaryWriter;
+        }
+
+        /// <summary>
+        /// Write a provided instance of BuildEventArgs to the BinaryWriter
+        /// </summary>
+        public void Write(BuildEventArgs e)
+        {
+            // the cases are ordered by most used first for performance
```

Comment:
I did quite extensive perf tuning for MSBuild Structured Log Viewer, so I have some stats on the number of most common events. Messages by far the most common. Also simple logic: there are more tasks than targets, more targets than projects, more projects than builds.